### ISSUE TITLE
Insert with Variant Selectors

### ISSUE THREAD

**Issue by fdw** [fdw/rofimoji#177]
I believe the variant selector is not actually copied into the application when entering. The emojis display correctly in rofi, but they don't get the selector once inserted.

_Originally posted by @mcourteaux in https://github.com/fdw/rofimoji/issues/142#issuecomment-2041386604_

**Comment by fdw** [comment]
Comments from [there](https://github.com/fdw/rofimoji/issues/142):

@fdw:
That is true, and it is on purpose: I didn't want rofimoji to try and interpret whether you wanted the Variant Selector or not. In my experience, it's usually correctly picked up by the target app anyway.

Do you think the Variant Selector should be always included?

---

@mcourteaux: 
> Do you think the Variant Selector should be always included?

Three options are reasonable to me:

    CLI option to say `--vs16=always`.
    CLI option to say `--vs16=never`.
    CLI option to say `--vs16=list-both`.

Where I'd make `--vs16=always` be the default-implied option. Maybe `--vs16` is a bit too obscure. Something more user-intuitive could be `--emoji-variant`.

Either way, I believe that what is shown in the options menu should also be the thing that gets actually inserted in the application. If rofi shows a colorful emoji, the inserted result should be a colorful emoji. If it's an old black-and-white symbol, so should be the inserted result.

---

@mcourteaux:
Or maybe: `--variants=15`, `--variants=16`, and `--variants=both`.
Or better: `--variant=bw`, `--variant=color`, and `--variant=both`.

**Comment by fdw** [comment]
I'm always hesitant to introduce parameters, as it adds complexity both in the code and for the user. Therefore, I'd like to understand your use case first: Which app do you use that can understand the variant but doesn't automatically add the variant or show a different glyph?

**Comment by mcourteaux** [comment]
> I'm always hesitant to introduce parameters, as it adds complexity both in the code and for the user.

As long as there is a sane default, I don't see the problem. The fix is very easy. You have a list of emojis which do contain explicitly the variant selector as their last character whenever they need one. Detecting that is easy, and stripping it as well. So handling the flag should be easy in general.

> Therefore, I'd like to understand your use case first: Which app do you use that can understand the variant but doesn't automatically add the variant or show a different glyph?

I'm just using the Kitty running nvim. I was writing a README earlier and wanted to add some emojis left and right, and half of them were black and white.

In Firefox, perhaps there is a specific problem with stopwatch, because when I try to insert that one, I get: ⏱ instead of :stopwatch:
However, trying "red heart", I get: ❤️, which is the emoji for me. So Firefox might do something special here to add the variant selector.
Another one "weights and scales" gives me: ⚖️ in Firefox, but ⚖ in Kitty.

**Comment by fdw** [comment]
> So handling the flag should be easy in general.

Please be a bit more mindful before telling someone that something is easy. There's a lot that's not visible on first glance; it's not just the code, but the maintainance - I will have to answer questions and feature requests for this thing, I will have to maintain it for the next years, making sure nothing breaks no matter what weird thing someone's trying to do.
So just the code might be easy (or it might not even be that), but that's just one factor among many.
Be nice to maintainers 🙂

Apart from that, though, I've just seen that `rofimoji` doesn't even try to strip the Variant Selector from the character. It's still there, just as you wanted, all the time. If Kitty shows you something else, that is more of a problem either there on your typer.

**Comment by mcourteaux** [comment]
> Please be a bit more mindful before telling someone that something is easy. There's a lot that's not visible on first glance; it's not just the code, but the maintainance - I will have to answer questions and feature requests for this thing, I will have to maintain it for the next years, making sure nothing breaks no matter what weird thing someone's trying to do.
> So just the code might be easy (or it might not even be that), but that's just one factor among many.
> Be nice to maintainers 🙂

I didn't mean to come across as disrespectful, but rather encouraging. I'm a serious programmer and I know what I'm talking about, so wanted to put in a few words of encouragement. And as you've shown, rofimoji is already doing what I was suggesting as the sane default, so I'm totally fine with not changing anything in the end.

> Apart from that, though, I've just seen that rofimoji doesn't even try to strip the Variant Selector from the character. It's still there, just as you wanted, all the time. If Kitty shows you something else, that is more of a problem either there on your typer.

I came to the same conclusion by looking at the code. I'm now investigating what the problem is. The problem seems to be the typer (xdotool). Opening nano, and inserting a bunch of times :scales: gets me this:

![image](https://github.com/fdw/rofimoji/assets/845012/b7c09471-fba5-4f24-baf9-b10d9b430b8c)

Very non-deterministic. Narrowing it down further, I created a file that contains only the emoji:

```
❯ echo -n "⚖<fe0f>" > emoji.test

~
❯ cat emoji.test
⚖️
~
❯ hexdump -C emoji.test
00000000  e2 9a 96 ef b8 8f                                 |......|
00000006

~
```

Then, repeatedly using `xdotool` to "type" the contents of this file, gives me:

```
❯ xdotool type --file emoji.test
⚖
~
❯ xdotool type --file emoji.test
⚖
~
❯ xdotool type --file emoji.test
⚖
~
❯ xdotool type --file emoji.test
⚖
~
❯ xdotool type --file emoji.test
⚖
~
❯ xdotool type --file emoji.test
⚖
~
❯ xdotool type --file emoji.test
⚖️
~
```
So either this is a xdotool bug, or this whole synthetic typing procedure is timing sensitive, and it therefore mostly fails.

From my experimentation, I conclude the following:

 - rofimoji is always doing the right thing.
 - kitty is always doing the right thing.
 - xdotool is not always doing the right thing.
 - zsh doesn't support correctly all emojis, in which case it prints the VS16 as `<fe0f>`.
 - bash does support the :balance_scale: emoji.
 - nano does support the :balance_scale: emoji.
 - nvim has a bug where sequences of :balance_scale: disappear.

![image](https://github.com/fdw/rofimoji/assets/845012/6cc76bb1-e75c-4fba-b42b-d40191dba312)

What a mess. Thanks for your input, it was helpful! And thanks for rofimoji. 🙏 (I had to try three times to get the thanks symbol to appear).

I'd analyze it with `xev`, but that's a bit overwhelming at this late hour of the day.

However, the good thing is, I believe I have an easy workaround. Of course, this is a thing that should get fixed in `xdotool`, but as I fear that is going to be a difficult task, I found this to be working: `xdotool type` the emoji codepoint, sleep a tiny bit, and then `xdotool type` the variation selector 16.

![image](https://github.com/fdw/rofimoji/assets/845012/fc9f9065-db8a-4ac7-a2d7-acc6dc74ab4d)

Implementing this workaround will be easy by just checking if the last 2 bytes are `fe0f`, and then splitting the xdotool command in two parts. I implemented this in #178.

**Comment by fdw** [comment]
Wow, thanks for the thorough analysis. That that look very convincing, and I'll look at your PR later (or the next days, sorry).

You can also have a look at [`xdotool`](https://github.com/jordansissel/xdotool)'s issues, maybe they can fix it.

PS:
> (I had to try three times to get the thanks symbol to appear).

At least for me, Firefox is often a problem with `xdotool`; everything else works better. I also have the impression that it gets worse under load, so maybe it is also a timing thing here?
